File: src/lib/societyBankAccount.ts
export type SocietyBankAccount = {
  id: string;
  society_id: string;
  bank_name: string;
  account_holder_name: string;
  account_number: string;
  ifsc: string;
  branch_name: string | null;
  branch_address: string | null;
  micr: string | null;
  account_type: string | null;
  currency: string;
  upi_vpa: string | null;
  customer_id: string | null;
  is_primary: boolean;
  is_active: boolean;
  effective_from: string | null;
  effective_to: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type SocietyBankAccountInput = {
  bank_name: string;
  account_holder_name: string;
  account_number: string;
  ifsc: string;
  branch_name?: string | null;
  branch_address?: string | null;
  micr?: string | null;
  account_type?: string | null;
  currency?: string;
  upi_vpa?: string | null;
  customer_id?: string | null;
  is_primary?: boolean;
  is_active?: boolean;
  effective_from?: string | null;
  effective_to?: string | null;
  notes?: string | null;
};

export function maskAccountNumber(accountNumber: string): string {
  const digits = accountNumber.replace(/\s+/g, '');
  if (digits.length <= 4) return digits;
  return `${'•'.repeat(Math.max(0, digits.length - 4))}${digits.slice(-4)}`;
}

export function buildUpiPayUrl(opts: {
  upiVpa: string;
  payeeName: string;
  amount?: number;
  note?: string;
}): string {
  const params = new URLSearchParams();
  params.set('pa', opts.upiVpa.trim());
  params.set('pn', opts.payeeName.trim() || 'Society');
  params.set('cu', 'INR');
  if (opts.amount && opts.amount > 0) params.set('am', String(opts.amount));
  if (opts.note?.trim()) params.set('tn', opts.note.trim());
  return `upi://pay?${params.toString()}`;
}

export function copyableBankDetails(account: Pick<
  SocietyBankAccount,
  'account_holder_name' | 'account_number' | 'ifsc' | 'bank_name' | 'branch_name' | 'upi_vpa'
>): string {
  const lines = [
    `Account name: ${account.account_holder_name}`,
    `Bank: ${account.bank_name}`,
    `Account number: ${account.account_number}`,
    `IFSC: ${account.ifsc}`,
  ];
  if (account.branch_name) lines.push(`Branch: ${account.branch_name}`);
  if (account.upi_vpa) lines.push(`UPI: ${account.upi_vpa}`);
  return lines.join('\n');
}

File: mobile/lib/models/society_bank_account.dart
class SocietyBankAccount {
  const SocietyBankAccount({
    required this.id,
    required this.societyId,
    required this.bankName,
    required this.accountHolderName,
    required this.accountNumber,
    required this.ifsc,
    this.branchName,
    this.upiVpa,
  });

  final String id;
  final String societyId;
  final String bankName;
  final String accountHolderName;
  final String accountNumber;
  final String ifsc;
  final String? branchName;
  final String? upiVpa;

  factory SocietyBankAccount.fromRow(Map<String, dynamic> row) => SocietyBankAccount(
        id: row['id'] as String,
        societyId: row['society_id'] as String,
        bankName: row['bank_name'] as String,
        accountHolderName: row['account_holder_name'] as String,
        accountNumber: row['account_number'] as String,
        ifsc: row['ifsc'] as String,
        branchName: row['branch_name'] as String?,
        upiVpa: row['upi_vpa'] as String?,
      );

  String get copyableDetails {
    final lines = <String>[
      'Account name: $accountHolderName',
      'Bank: $bankName',
      'Account number: $accountNumber',
      'IFSC: $ifsc',
    ];
    if (branchName != null && branchName!.trim().isNotEmpty) {
      lines.add('Branch: $branchName');
    }
    if (upiVpa != null && upiVpa!.trim().isNotEmpty) {
      lines.add('UPI: $upiVpa');
    }
    return lines.join('\n');
  }
}
